library/core/src/fmt/num.rs

```diff
@@ -229,58 +265,64 @@ macro_rules! impl_Display {
             // is safe to access.
             unsafe {
                 // need at least 16 bits for the 4-characters-at-a-time to work.
-                assert!(crate::mem::size_of::<$u>() >= 2);
-
-                // eagerly decode 4 characters at a time
-                while n >= 10000 {
-                    let rem = (n % 10000) as usize;
-                    n /= 10000;
-
-                    let d1 = (rem / 100) << 1;
-                    let d2 = (rem % 100) << 1;
-                    curr -= 4;
-
-                    // We are allowed to copy to `buf_ptr[curr..curr + 3]` here since
-                    // otherwise `curr < 0`. But then `n` was originally at least `10000^10`
-                    // which is `10^40 > 2^128 > n`.
-                    ptr::copy_nonoverlapping(lut_ptr.add(d1), buf_ptr.add(curr), 2);
-                    ptr::copy_nonoverlapping(lut_ptr.add(d2), buf_ptr.add(curr + 2), 2);
+                #[allow(overflowing_literals)]
+                #[allow(unused_comparisons)]
+                // This block should be removed for smaller types at compile time so it
+                // should be ok.
+                if core::mem::size_of::<$t>() >= 2 {
+                    // eagerly decode 4 characters at a time
```

Comment:
I actually inspected this in Godbolt, and if we *remove* this `while`, then compile this function for a u8, and then immediately up-size it to u16 (which makes the literals inside this valid)...

...nothing happens, under optimizations.

The `while`'s loop-condition is correctly recognized as impossible and removed.

Please do not introduce doubling of conditionals, they just make code harder to read and make it unclear what actually controls the logic.

Reply:
Fine by me!

Reply:
Let's maybe remove it in a follow-up to keep the diff easier to track. ^^'

Reply:
You changed an assert to an `if`.